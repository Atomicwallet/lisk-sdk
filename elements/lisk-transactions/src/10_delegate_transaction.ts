/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { BaseTransaction, StateStore } from './base_transaction';
import { CHAIN_STATE_DELEGATE_USERNAMES, DELEGATE_NAME_FEE } from './constants';
import { TransactionError } from './errors';
import { BaseTransactionInput } from './types';

interface RegisteredDelegate {
	readonly username: string;
	readonly address: Buffer;
}
interface ChainUsernames {
	readonly registeredDelegates: RegisteredDelegate[];
}

export interface DelegateAsset {
	readonly username: string;
}

export const delegateRegistrationAssetSchema = {
	$id: 'lisk/delegate-registration-transaction',
	type: 'object',
	required: ['username'],
	properties: {
		username: {
			dataType: 'string',
			fieldNumber: 1,
			minLength: 1,
			maxLength: 20,
		},
	},
};

export class DelegateTransaction extends BaseTransaction {
	public static TYPE = 10;
	public static NAME_FEE = BigInt(DELEGATE_NAME_FEE);
	public static ASSET_SCHEMA = delegateRegistrationAssetSchema;
	public readonly asset: DelegateAsset;

	public constructor(transaction: BaseTransactionInput<DelegateAsset>) {
		super(transaction);

		this.asset = transaction.asset;
	}

	protected async applyAsset(
		store: StateStore,
	): Promise<ReadonlyArray<TransactionError>> {
		const errors: TransactionError[] = [];
		const sender = await store.account.get(this.senderId);

		// Data format for the registered delegates
		// chain:delegateUsernames => { registeredDelegates: { username, address }[] }
		const usernamesBuffer = await store.chain.get(
			CHAIN_STATE_DELEGATE_USERNAMES,
		);
		const usernames = usernamesBuffer
			? (JSON.parse(usernamesBuffer.toString('utf8')) as ChainUsernames)
			: { registeredDelegates: [] };
		const usernameExists = usernames.registeredDelegates.find(
			delegate => delegate.username === this.asset.username,
		);

		if (!usernameExists) {
			usernames.registeredDelegates.push({
				username: this.asset.username,
				address: this.senderId,
			});
			usernames.registeredDelegates.sort((a, b) =>
				a.address.compare(b.address),
			);
			store.chain.set(
				CHAIN_STATE_DELEGATE_USERNAMES,
				Buffer.from(JSON.stringify(usernames), 'utf8'),
			);
		}

		if (usernameExists) {
			errors.push(
				new TransactionError(
					`Username is not unique.`,
					this.id,
					'.asset.username',
				),
			);
		}
		if (sender.isDelegate || sender.username) {
			errors.push(
				new TransactionError(
					'Account is already a delegate',
					this.id,
					'.asset.username',
				),
			);
		}
		sender.username = this.asset.username;
		sender.isDelegate = 1;
		store.account.set(sender.address, sender);

		return errors;
	}

	protected async undoAsset(
		store: StateStore,
	): Promise<ReadonlyArray<TransactionError>> {
		const sender = await store.account.get(this.senderId);

		// Data format for the registered delegates
		// chain:delegateUsernames => { registeredDelegates: { username, address }[] }
		const usernamesBuffer = await store.chain.get(
			CHAIN_STATE_DELEGATE_USERNAMES,
		);
		const usernames = usernamesBuffer
			? (JSON.parse(usernamesBuffer.toString('utf8')) as ChainUsernames)
			: { registeredDelegates: [] };
		const updatedRegisteredDelegates = {
			registeredDelegates: usernames.registeredDelegates.filter(
				delegate => delegate.username !== sender.username,
			),
		};
		updatedRegisteredDelegates.registeredDelegates.sort((a, b) =>
			a.address.compare(b.address),
		);
		store.chain.set(
			CHAIN_STATE_DELEGATE_USERNAMES,
			Buffer.from(JSON.stringify(updatedRegisteredDelegates), 'utf8'),
		);

		sender.username = null;
		sender.isDelegate = 0;
		store.account.set(sender.address, sender);
		return [];
	}
}
