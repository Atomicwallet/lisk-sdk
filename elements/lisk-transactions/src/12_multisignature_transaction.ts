/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import {
	getAddressAndPublicKeyFromPassphrase,
	getAddressFromPublicKey,
	signData,
	bufferToHex,
	hash,
} from '@liskhq/lisk-cryptography';

import { BaseTransaction, StateStore } from './base_transaction';
import { TransactionError } from './errors';
import { createResponse, TransactionResponse } from './response';
import {
	buildPublicKeyPassphraseDict,
	sortKeysAscending,
	validateKeysSignatures,
	validateSignature,
} from './utils';
import { BaseTransactionInput, AccountAsset } from './types';

export const multisigRegAssetSchema = {
	$id: 'lisk/multisignature-registration-transaction',
	type: 'object',
	required: ['numberOfSignatures', 'optionalKeys', 'mandatoryKeys'],
	properties: {
		numberOfSignatures: {
			dataType: 'uint32',
			fieldNumber: 1,
			minimum: 1,
			maximum: 64,
		},
		mandatoryKeys: {
			type: 'array',
			items: {
				dataType: 'bytes',
			},
			fieldNumber: 2,
			minItems: 0,
			maxItems: 64,
			minLength: 32,
			maxLength: 32,
		},
		optionalKeys: {
			type: 'array',
			items: {
				dataType: 'bytes',
			},
			fieldNumber: 3,
			minItems: 0,
			maxItems: 64,
			minLength: 32,
			maxLength: 32,
		},
	},
};

const setMemberAccounts = async (
	store: StateStore,
	membersPublicKeys: Array<Readonly<Buffer>>,
): Promise<void> => {
	for (const memberPublicKey of membersPublicKeys) {
		const address = getAddressFromPublicKey(memberPublicKey as Buffer);
		// Key might not exists in the blockchain yet so we fetch or default
		const memberAccount = await store.account.getOrDefault(address);
		store.account.set(memberAccount.address, memberAccount);
	}
};

const hasDuplicateKey = (keys: Buffer[]): boolean => {
	const temp: { [key: string]: boolean | undefined } = {};
	for (const key of keys) {
		if (temp[key.toString('base64')]) {
			return true;
		}
		temp[key.toString('base64')] = true;
	}
	return false;
};

export interface MultiSignatureAsset {
	mandatoryKeys: Array<Readonly<Buffer>>;
	optionalKeys: Array<Readonly<Buffer>>;
	readonly numberOfSignatures: number;
}

export class MultisignatureTransaction extends BaseTransaction {
	public static TYPE = 12;
	public static ASSET_SCHEMA = multisigRegAssetSchema;
	public readonly asset: MultiSignatureAsset;
	private readonly MAX_KEYS_COUNT = 64;

	public constructor(transaction: BaseTransactionInput<MultiSignatureAsset>) {
		super(transaction);

		this.asset = transaction.asset;
	}

	// Verifies multisig signatures as per LIP-0017
	// eslint-disable-next-line @typescript-eslint/require-await
	public async verifySignatures(
		store: StateStore,
	): Promise<TransactionResponse> {
		const { networkIdentifier } = store.chain;
		const transactionBytes = this.getSigningBytes();
		const transactionWithNetworkIdentifierBytes = Buffer.concat([
			networkIdentifier,
			transactionBytes,
		]);

		const { mandatoryKeys, optionalKeys } = this.asset;

		// For multisig registration we need all signatures to be present
		if (
			mandatoryKeys.length + optionalKeys.length + 1 !==
			this.signatures.length
		) {
			return createResponse(this.id, [
				new TransactionError('There are missing signatures'),
			]);
		}

		// Check if empty signatures are present
		if (!this.signatures.every(signature => signature.length > 0)) {
			return createResponse(this.id, [
				new TransactionError(
					'A signature is required for each registered key.',
				),
			]);
		}

		// Verify first signature is from senderPublicKey
		const { valid, error } = validateSignature(
			this.senderPublicKey,
			this.signatures[0] as Buffer,
			transactionWithNetworkIdentifierBytes,
		);

		if (!valid) {
			return createResponse(this.id, [error as TransactionError]);
		}

		// Verify each mandatory key signed in order
		const mandatorySignaturesErrors = validateKeysSignatures(
			mandatoryKeys,
			this.signatures.slice(1, mandatoryKeys.length + 1),
			transactionWithNetworkIdentifierBytes,
		);

		if (mandatorySignaturesErrors.length) {
			return createResponse(this.id, mandatorySignaturesErrors);
		}
		// Verify each optional key signed in order
		const optionalSignaturesErrors = validateKeysSignatures(
			optionalKeys,
			this.signatures.slice(mandatoryKeys.length + 1),
			transactionWithNetworkIdentifierBytes,
		);
		if (optionalSignaturesErrors.length) {
			return createResponse(this.id, optionalSignaturesErrors);
		}

		return createResponse(this.id, []);
	}

	public sign(
		networkIdentifier: Buffer,
		senderPassphrase: string,
		passphrases?: ReadonlyArray<string>,
		keys?: {
			readonly mandatoryKeys: Array<Readonly<Buffer>>;
			readonly optionalKeys: Array<Readonly<Buffer>>;
			readonly numberOfSignatures: number;
		},
	): void {
		// Sort the keys in the transaction
		sortKeysAscending(this.asset.mandatoryKeys);
		sortKeysAscending(this.asset.optionalKeys);

		// Sign with sender
		const { publicKey } = getAddressAndPublicKeyFromPassphrase(
			senderPassphrase,
		);

		if (!this.senderPublicKey.equals(publicKey)) {
			throw new Error(
				'Transaction senderPublicKey does not match public key from passphrase',
			);
		}

		this.senderPublicKey = publicKey;

		const transactionWithNetworkIdentifierBytes = Buffer.concat([
			networkIdentifier,
			this.getSigningBytes(),
		]);

		this.signatures.push(
			signData(transactionWithNetworkIdentifierBytes, senderPassphrase),
		);

		// Sign with members
		if (keys && passphrases) {
			const keysAndPassphrases = buildPublicKeyPassphraseDict([...passphrases]);
			// Make sure passed in keys are sorted
			sortKeysAscending(keys.mandatoryKeys);
			sortKeysAscending(keys.optionalKeys);
			// Sign with all keys
			for (const aKey of [...keys.mandatoryKeys, ...keys.optionalKeys]) {
				const senderPublicKeyStr = bufferToHex(aKey as Buffer);
				// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
				if (keysAndPassphrases[senderPublicKeyStr]) {
					const { passphrase } = keysAndPassphrases[senderPublicKeyStr];
					this.signatures.push(
						signData(transactionWithNetworkIdentifierBytes, passphrase),
					);
				} else {
					// Push an empty signature if a passphrase is missing
					this.signatures.push(Buffer.from(''));
				}
			}
		}
		this._id = hash(this.getBytes());
	}

	protected validateAsset(): ReadonlyArray<TransactionError> {
		const errors = [];

		const { mandatoryKeys, optionalKeys, numberOfSignatures } = this.asset;

		if (hasDuplicateKey(mandatoryKeys as Buffer[])) {
			errors.push(
				new TransactionError(
					'MandatoryKeys contains duplicate public keys',
					this.id,
					'.asset.mandatoryKeys',
				),
			);
		}

		if (hasDuplicateKey(optionalKeys as Buffer[])) {
			errors.push(
				new TransactionError(
					'OptionalKeys contains duplicate public keys',
					this.id,
					'.asset.optionalKeys',
				),
			);
		}

		// Check if key count is less than number of required signatures
		if (mandatoryKeys.length + optionalKeys.length < numberOfSignatures) {
			errors.push(
				new TransactionError(
					'The numberOfSignatures is bigger than the count of Mandatory and Optional keys',
					this.id,
					'.asset.numberOfSignatures',
					this.asset.numberOfSignatures,
				),
			);
		}

		// Check if key count is less than 1
		if (
			mandatoryKeys.length + optionalKeys.length > this.MAX_KEYS_COUNT ||
			mandatoryKeys.length + optionalKeys.length <= 0
		) {
			errors.push(
				new TransactionError(
					'The count of Mandatory and Optional keys should be between 1 and 64',
					this.id,
					'.asset.optionalKeys .asset.mandatoryKeys',
					this.asset.numberOfSignatures,
				),
			);
		}

		// The numberOfSignatures needs to be equal or bigger than number of mandatoryKeys
		if (mandatoryKeys.length > numberOfSignatures) {
			errors.push(
				new TransactionError(
					'The numberOfSignatures needs to be equal or bigger than the number of Mandatory keys',
					this.id,
					'.asset.numberOfSignatures',
					this.asset.numberOfSignatures,
				),
			);
		}

		if (errors.length > 0) {
			return errors;
		}

		// Check if keys are repeated between mandatory and optional key sets
		const repeatedKeys = mandatoryKeys.filter(
			value =>
				optionalKeys.find(optional => optional.equals(value as Buffer)) !==
				undefined,
		);
		if (repeatedKeys.length > 0) {
			errors.push(
				new TransactionError(
					'Invalid combination of Mandatory and Optional keys',
					this.id,
					'.asset.mandatoryKeys, .asset.optionalKeys',
					repeatedKeys.join(', '),
				),
			);
		}

		if (errors.length > 0) {
			return errors;
		}

		// Check if the length of mandatory, optional and sender keys matches the length of signatures
		if (
			mandatoryKeys.length + optionalKeys.length + 1 !==
			this.signatures.length
		) {
			return [
				new TransactionError(
					'The number of mandatory, optional and sender keys should match the number of signatures',
					this.id,
				),
			];
		}

		// Check keys are sorted lexicographically
		const sortedMandatoryKeys = [...mandatoryKeys].sort((a, b) =>
			a.compare(b as Buffer),
		);
		const sortedOptionalKeys = [...optionalKeys].sort((a, b) =>
			a.compare(b as Buffer),
		);
		for (let i = 0; i < sortedMandatoryKeys.length; i += 1) {
			if (!mandatoryKeys[i].equals(sortedMandatoryKeys[i] as Buffer)) {
				errors.push(
					new TransactionError(
						'Mandatory keys should be sorted lexicographically',
						this.id,
						'.asset.mandatoryKeys',
						mandatoryKeys.join(', '),
					),
				);
				break;
			}
		}

		for (let i = 0; i < sortedOptionalKeys.length; i += 1) {
			if (!optionalKeys[i].equals(sortedOptionalKeys[i] as Buffer)) {
				errors.push(
					new TransactionError(
						'Optional keys should be sorted lexicographically',
						this.id,
						'.asset.optionalKeys',
						optionalKeys.join(', '),
					),
				);
				break;
			}
		}

		return errors;
	}

	protected async applyAsset(
		store: StateStore,
	): Promise<ReadonlyArray<TransactionError>> {
		const errors: TransactionError[] = [];
		const sender = await store.account.get<AccountAsset>(this.senderId);

		// Check if multisignatures already exists on account
		if (sender.keys.numberOfSignatures > 0) {
			errors.push(
				new TransactionError(
					'Register multisignature only allowed once per account.',
					this.id,
					'.signatures',
				),
			);
		}

		sender.keys = {
			numberOfSignatures: this.asset.numberOfSignatures,
			mandatoryKeys: this.asset.mandatoryKeys as Buffer[],
			optionalKeys: this.asset.optionalKeys as Buffer[],
		};

		store.account.set(sender.address, sender);

		// Cache all members public keys
		await setMemberAccounts(store, sender.keys.mandatoryKeys);
		await setMemberAccounts(store, sender.keys.optionalKeys);

		return errors;
	}
}
