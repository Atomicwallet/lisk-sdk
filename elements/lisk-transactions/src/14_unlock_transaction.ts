/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { BaseTransaction, StateStore } from './base_transaction';
import { TransactionError } from './errors';
import { Account, BaseTransactionInput } from './types';
import { getPunishmentPeriod, sortUnlocking } from './utils';

export interface Unlock {
	readonly delegateAddress: Buffer;
	readonly amount: bigint;
	readonly unvoteHeight: number;
}

export interface UnlockAsset {
	readonly unlockObjects: ReadonlyArray<Unlock>;
}

const unlockAssetSchema = {
	$id: 'lisk/unlock-transaction',
	type: 'object',
	required: ['unlockObjects'],
	properties: {
		unlockObjects: {
			type: 'array',
			items: {
				type: 'object',
				required: ['delegateAddress', 'amount', 'unvoteHeight'],
				properties: {
					delegateAddress: {
						dataType: 'bytes',
						fieldNumber: 1,
						minLength: 20,
						maxLength: 20,
					},
					amount: {
						dataType: 'uint64',
						fieldNumber: 2,
					},
					unvoteHeight: {
						dataType: 'uint32',
						fieldNumber: 3,
					},
				},
			},
			fieldNumber: 1,
		},
	},
};

const AMOUNT_MULTIPLIER_FOR_VOTES = BigInt(10) * BigInt(10) ** BigInt(8);
const WAIT_TIME_VOTE = 2000;
const WAIT_TIME_SELF_VOTE = 260000;

const getWaitingPeriod = (
	sender: Account,
	delegateAccount: Account,
	lastBlockHeight: number,
	unlockObject: Unlock,
): number => {
	const currentHeight = lastBlockHeight + 1;
	const waitTime =
		sender.address === delegateAccount.address
			? WAIT_TIME_SELF_VOTE
			: WAIT_TIME_VOTE;

	return waitTime - (currentHeight - unlockObject.unvoteHeight);
};

export class UnlockTransaction extends BaseTransaction {
	public static TYPE = 14;
	public static ASSET_SCHEMA = unlockAssetSchema;
	public readonly asset: UnlockAsset;

	public constructor(transaction: BaseTransactionInput<UnlockAsset>) {
		super(transaction);

		this.asset = transaction.asset;
	}

	protected validateAsset(): ReadonlyArray<TransactionError> {
		const errors = [];

		for (const unlock of this.asset.unlockObjects) {
			if (unlock.amount <= BigInt(0)) {
				errors.push(
					new TransactionError(
						'Amount cannot be less than or equal to zero',
						this.id,
						'.asset.votes.amount',
						unlock.amount.toString(),
					),
				);
			}

			if (unlock.amount % AMOUNT_MULTIPLIER_FOR_VOTES !== BigInt(0)) {
				errors.push(
					new TransactionError(
						'Amount should be multiple of 10 * 10^8',
						this.id,
						'.asset.votes.amount',
						unlock.amount.toString(),
					),
				);
			}
		}

		return errors;
	}

	protected async applyAsset(
		store: StateStore,
	): Promise<ReadonlyArray<TransactionError>> {
		const errors = [];

		for (const unlock of this.asset.unlockObjects) {
			const sender = await store.account.get(this.senderId);
			const delegate = await store.account.getOrDefault(unlock.delegateAddress);
			if (!delegate.username) {
				errors.push(
					new TransactionError(
						'Voted account is not registered as delegate',
						this.id,
						'.asset.unlockObjects.delegateAddress',
					),
				);
				// eslint-disable-next-line no-continue
				continue;
			}

			const waitingPeriod = getWaitingPeriod(
				sender,
				delegate,
				store.chain.lastBlockHeader.height,
				unlock,
			);
			if (waitingPeriod > 0) {
				errors.push(
					new TransactionError(
						'Unlocking is not permitted as it is still within the waiting period',
						this.id,
						'.asset.unlockObjects.unvoteHeight',
						waitingPeriod,
						0,
					),
				);
			}
			const punishmentPeriod = getPunishmentPeriod(
				sender,
				delegate,
				store.chain.lastBlockHeader.height,
			);
			if (punishmentPeriod > 0) {
				errors.push(
					new TransactionError(
						'Unlocking is not permitted as delegate is currently being punished',
						this.id,
						'.asset.unlockObjects.delegateAddress',
						punishmentPeriod,
						0,
					),
				);
			}
			const unlockIndex = sender.unlocking.findIndex(
				obj =>
					obj.amount === unlock.amount &&
					obj.delegateAddress === unlock.delegateAddress &&
					obj.unvoteHeight === unlock.unvoteHeight,
			);
			if (unlockIndex < 0) {
				errors.push(
					new TransactionError(
						'Corresponding unlocking object not found',
						this.id,
						'.asset.unlockObjects',
					),
				);
				// eslint-disable-next-line no-continue
				continue;
			}
			sender.unlocking.splice(unlockIndex, 1);
			sender.balance += unlock.amount;
			store.account.set(sender.address, sender);
		}

		return errors;
	}

	protected async undoAsset(
		store: StateStore,
	): Promise<ReadonlyArray<TransactionError>> {
		for (const unlock of this.asset.unlockObjects) {
			const sender = await store.account.get(this.senderId);

			sender.balance -= unlock.amount;
			sender.unlocking.push(unlock);
			// Resort the unlocking since it's pushed to the end
			sortUnlocking(sender.unlocking);
			store.account.set(sender.address, sender);
		}

		return [];
	}
}
