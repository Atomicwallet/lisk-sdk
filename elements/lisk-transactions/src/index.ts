/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { DelegateTransaction } from './10_delegate_transaction';
import { MultisignatureTransaction } from './12_multisignature_transaction';
import { VoteTransaction } from './13_vote_transaction';
import { UnlockTransaction } from './14_unlock_transaction';
import { ProofOfMisbehaviorTransaction } from './15_proof_of_misbehavior_transaction';
import { TransferTransaction } from './8_transfer_transaction';
import { BaseTransaction, StateStore } from './base_transaction';
import * as constants from './constants';
import { convertToAssetError, convertToTransactionError, TransactionError } from './errors';
import { createResponse, Status, TransactionResponse } from './response';
import { transactionInterface } from './schema';
import {
	getSigningBytes,
	validateTransactionSchema,
	signTransaction,
	signMultiSignatureTransaction,
} from './transaction_helper';
import { Account, TransactionJSON } from './types';
import {
	convertBeddowsToLSK,
	convertLSKToBeddows,
	prependMinusToPublicKeys,
	prependPlusToPublicKeys,
	validateSenderIdAndPublicKey,
	validateSignature,
	verifyMinRemainingBalance,
	verifyMultiSignatureTransaction,
} from './utils';

const exposedUtils = {
	convertBeddowsToLSK,
	convertLSKToBeddows,
	prependMinusToPublicKeys,
	prependPlusToPublicKeys,
	validateSignature,
	verifyMinRemainingBalance,
	validateSenderIdAndPublicKey,
	verifyMultiSignatureTransaction,
};

export {
	Account,
	BaseTransaction,
	getSigningBytes,
	StateStore,
	TransferTransaction,
	DelegateTransaction,
	VoteTransaction,
	MultisignatureTransaction,
	UnlockTransaction,
	createResponse,
	ProofOfMisbehaviorTransaction,
	signTransaction,
	signMultiSignatureTransaction,
	Status,
	TransactionResponse,
	TransactionJSON,
	TransactionError,
	validateTransactionSchema,
	transactionInterface,
	convertToAssetError,
	convertToTransactionError,
	constants,
	exposedUtils as utils,
};
