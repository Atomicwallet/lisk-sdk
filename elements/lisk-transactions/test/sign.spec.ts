/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { getAddressAndPublicKeyFromPassphrase } from '@liskhq/lisk-cryptography';
import { getSigningBytes, signTransaction, signMultiSignatureTransaction } from '../src/sign';

const validTransaction = {
	type: 8,
	nonce: BigInt('1'),
	fee: BigInt('10000000'),
	senderPublicKey: Buffer.from(
		'0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a',
		'hex',
	),
	asset: {
		recipientAddress: Buffer.from('3a971fd02b4a07fc20aad1936d3cb1d263b96e0f', 'hex'),
		amount: BigInt('4008489300000000'),
		data: '',
	},
};

const validAssetSchema = {
	$id: 'lisk/transfer-transaction',
	title: 'Transfer transaction asset',
	type: 'object',
	required: ['amount', 'recipientAddress', 'data'],
	properties: {
		amount: {
			dataType: 'uint64',
			fieldNumber: 1,
		},
		recipientAddress: {
			dataType: 'bytes',
			fieldNumber: 2,
			minLength: 20,
			maxLength: 20,
		},
		data: {
			dataType: 'string',
			fieldNumber: 3,
			minLength: 0,
			maxLength: 64,
		},
	},
};
const networkIdentifier = Buffer.from(
	'e48feb88db5b5cf5ad71d93cdcd1d879b6d5ed187a36b0002cc34e0ef9883255',
	'base64',
);
const passphrase1 = 'trim elegant oven term access apple obtain error grain excite lawn neck';
const passphrase2 =
	'sugar object slender confirm clock peanut auto spice carbon knife increase estate';
const passphrase3 = 'faculty inspire crouch quit sorry vague hard ski scrap jaguar garment limb';
const { publicKey: publicKey1 } = getAddressAndPublicKeyFromPassphrase(passphrase1);
const { publicKey: publicKey2 } = getAddressAndPublicKeyFromPassphrase(passphrase2);
const { publicKey: publicKey3 } = getAddressAndPublicKeyFromPassphrase(passphrase3);
const keys = {
	mandatoryKeys: [publicKey1, publicKey2],
	optionalKeys: [publicKey3],
};

describe('getSigningBytes', () => {
	it('should throw error for invalid transaction object', () => {
		const invalidTransactionObjects = [
			{ ...validTransaction, type: BigInt(8) },
			{ ...validTransaction, nonce: 1 },
			{ ...validTransaction, fee: 1000000 },
			{ ...validTransaction, senderPublicKey: 1 },
		];
		return invalidTransactionObjects.forEach(transactionObject =>
			expect(() => getSigningBytes(validAssetSchema, transactionObject)).toThrow(),
		);
	});

	it('should throw error when asset is null', () => {
		return expect(() =>
			getSigningBytes(validAssetSchema, { ...validTransaction, asset: null }),
		).toThrow(new Error('Transaction object asset must be of type object and not null'));
	});

	it('should throw error for invalid asset object', () => {
		const invalidAssets = [
			{ ...validTransaction, asset: { ...validTransaction.asset, amount: 1000 } },
			{
				...validTransaction,
				asset: { ...validTransaction.asset, recipientAddress: 'dummyAddress' },
			},
		];
		return invalidAssets.forEach(transactionObject =>
			expect(() => getSigningBytes(validAssetSchema, transactionObject)).toThrow(),
		);
	});
});

describe('signTransaction', () => {
	it('should throw error for invalid transaction object', () => {
		const invalidTransactionObjects = [
			{ ...validTransaction, type: BigInt(8) },
			{ ...validTransaction, nonce: 1 },
			{ ...validTransaction, fee: 1000000 },
			{ ...validTransaction, senderPublicKey: 1 },
		];
		return invalidTransactionObjects.forEach(transactionObject =>
			expect(() =>
				signTransaction(validAssetSchema, transactionObject, networkIdentifier, passphrase1),
			).toThrow(),
		);
	});

	it('should throw error when asset is null', () => {
		return expect(() =>
			signTransaction(
				validAssetSchema,
				{ ...validTransaction, asset: null },
				networkIdentifier,
				passphrase1,
			),
		).toThrow(new Error('Transaction object asset must be of type object and not null'));
	});

	it('should throw error for invalid asset object', () => {
		const invalidAssets = [
			{ ...validTransaction, asset: { ...validTransaction.asset, amount: 1000 } },
			{
				...validTransaction,
				asset: { ...validTransaction.asset, recipientAddress: 'dummyAddress' },
			},
		];
		return invalidAssets.forEach(transactionObject =>
			expect(() =>
				signTransaction(validAssetSchema, transactionObject, networkIdentifier, passphrase1),
			).toThrow(),
		);
	});
});

describe('signMultiSignatureTransaction', () => {
	it('should throw error for invalid transaction object', () => {
		const invalidTransactionObjects = [
			{ ...validTransaction, type: BigInt(8) },
			{ ...validTransaction, nonce: 1 },
			{ ...validTransaction, fee: 1000000 },
			{ ...validTransaction, senderPublicKey: 1 },
		];
		return invalidTransactionObjects.forEach(transactionObject =>
			expect(() =>
				signMultiSignatureTransaction(
					validAssetSchema,
					transactionObject,
					networkIdentifier,
					passphrase1,
					keys,
				),
			).toThrow(),
		);
	});

	it('should throw error when asset is null', () => {
		return expect(() =>
			signMultiSignatureTransaction(
				validAssetSchema,
				{ ...validTransaction, asset: null },
				networkIdentifier,
				passphrase1,
				keys,
			),
		).toThrow(new Error('Transaction object asset must be of type object and not null'));
	});

	it('should throw error for invalid asset object', () => {
		const invalidAssets = [
			{ ...validTransaction, asset: { ...validTransaction.asset, amount: 1000 } },
			{
				...validTransaction,
				asset: { ...validTransaction.asset, recipientAddress: 'dummyAddress' },
			},
		];
		return invalidAssets.forEach(transactionObject =>
			expect(() =>
				signMultiSignatureTransaction(
					validAssetSchema,
					transactionObject,
					networkIdentifier,
					passphrase1,
					keys,
				),
			).toThrow(),
		);
	});
});
