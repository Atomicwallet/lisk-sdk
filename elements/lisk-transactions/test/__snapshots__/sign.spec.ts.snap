// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sign getSigningBytes should return transaction bytes for given asset 1`] = `
Object {
  "data": Array [
    8,
    2,
    16,
    0,
    24,
    1,
    32,
    128,
    173,
    226,
    4,
    42,
    32,
    241,
    185,
    244,
    238,
    113,
    181,
    213,
    133,
    125,
    59,
    52,
    109,
    68,
    28,
    169,
    103,
    242,
    120,
    112,
    235,
    238,
    136,
    86,
    157,
    179,
    100,
    253,
    19,
    226,
    138,
    219,
    163,
    50,
    33,
    8,
    128,
    250,
    191,
    166,
    178,
    182,
    143,
    7,
    18,
    20,
    58,
    151,
    31,
    208,
    43,
    74,
    7,
    252,
    32,
    170,
    209,
    147,
    109,
    60,
    177,
    210,
    99,
    185,
    110,
    15,
    26,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`sign signTransaction should return signed transaction for given asset schema 1`] = `
Object {
  "asset": Object {
    "amount": 4008489300000000n,
    "data": "",
    "recipientAddress": Object {
      "data": Array [
        58,
        151,
        31,
        208,
        43,
        74,
        7,
        252,
        32,
        170,
        209,
        147,
        109,
        60,
        177,
        210,
        99,
        185,
        110,
        15,
      ],
      "type": "Buffer",
    },
  },
  "assetID": 0,
  "fee": 10000000n,
  "id": Object {
    "data": Array [
      45,
      149,
      204,
      201,
      194,
      80,
      188,
      27,
      88,
      109,
      194,
      94,
      89,
      251,
      163,
      245,
      188,
      221,
      255,
      240,
      92,
      167,
      134,
      83,
      188,
      156,
      197,
      67,
      64,
      211,
      212,
      189,
    ],
    "type": "Buffer",
  },
  "moduleID": 2,
  "nonce": 1n,
  "senderPublicKey": Object {
    "data": Array [
      241,
      185,
      244,
      238,
      113,
      181,
      213,
      133,
      125,
      59,
      52,
      109,
      68,
      28,
      169,
      103,
      242,
      120,
      112,
      235,
      238,
      136,
      86,
      157,
      179,
      100,
      253,
      19,
      226,
      138,
      219,
      163,
    ],
    "type": "Buffer",
  },
  "signatures": Array [
    Object {
      "data": Array [
        238,
        171,
        74,
        24,
        224,
        198,
        160,
        170,
        147,
        239,
        3,
        241,
        66,
        81,
        87,
        33,
        131,
        124,
        154,
        16,
        54,
        158,
        191,
        62,
        162,
        136,
        58,
        167,
        72,
        220,
        215,
        32,
        3,
        59,
        164,
        235,
        249,
        252,
        222,
        209,
        11,
        227,
        174,
        70,
        42,
        224,
        159,
        110,
        147,
        61,
        75,
        189,
        237,
        235,
        212,
        255,
        50,
        111,
        121,
        26,
        247,
        220,
        63,
        11,
      ],
      "type": "Buffer",
    },
  ],
}
`;

exports[`sign when running scenario "multisignature_registration_transaction" Both mandatory and optional member group should have correct signatures 1`] = `[Function]`;

exports[`sign when running scenario "multisignature_registration_transaction" Only mandatory members should have correct signatures 1`] = `[Function]`;

exports[`sign when running scenario "multisignature_registration_transaction" Only optional members should have correct signatures 1`] = `[Function]`;

exports[`sign when running scenario "multisignature_registration_transaction" Second signature case should have correct signatures 1`] = `[Function]`;

exports[`sign when running scenario "multisignature_registration_transaction" Sender is a member of the group should have correct signatures 1`] = `[Function]`;
